### ISSUE TITLE
View Coming Soon titles as in app

### ISSUE THREAD

**Issue by tarhan** [CastagnaIT/plugin.video.netflix#1117]
## Feature request
I'm submitting a proposal to ...
  - [* ] new feature
  - [ ] change an existing feature

### Describe your request

Within Netflix app I can view "Coming Soon" titles and request notification alerts for some of them. Is it possible to add at least viewing such list?  

#### Development tips

I saw within #107 issue example "model" comes from Netflix I think. Within that JSON document there are keys like `hasComingSoon`, `hasComingSoonFeed` and `netflix.ui.coming.soon.feed.enabled`. So I think there way to "just" enable Comming Soon feed.

**Comment by CastagnaIT** [comment]
> Is it possible to add at least viewing such list?

The mobile app "Coming Soon" is a bit different from the website where we are based, i do not know what query do the mobile app maybe a reverse engineering could show the query? and get the list in same way, if this is a single query should be simple to add it

instead the website have the "New & Popular" section with more subsections,
to add this ("New & Popular") menu is already planned on Projects page but this require to study it and different works to do,
and ATM i dont think i will

> I saw within #107 issue example "model" comes from Netflix I think...

these settings come from website backend, can not be changed only read
unfortunately these not help

**Comment by tarhan** [comment]
>  maybe a reverse engineering could show the query?
Can you tell if request for other section done with MSL? I've not read yet spec for MSL. 
Other section list and content requests encrypted or just signed?
> and ATM i dont think i will
Do you mean that other person will implement "New & Popular" section?

**Comment by CastagnaIT** [comment]
> Can you tell if request for other section done with MSL? I've not read yet spec for MSL.
Other section list and content requests encrypted or just signed?

these type of requests are done on standard HTTPS without MSL

> Do you mean that other person will implement "New & Popular" section?

could be but ATM i am the only dev

**Comment by CastagnaIT** [comment]
thanks this is perfect!
what software you have used to catch the mobile https traffic?
i have tried more time but without success i do not know where i am wrong or if it is my smartphone too old...

yes this format type is "jsonGraph" a proprietary netflix format type
i will try to see if these path are working on our endpoint

sidenote: please next time do not paste entire code in to comment,
instead attach a file or make a link to the content

**Comment by tarhan** [comment]
> what software you have used to catch the mobile https traffic?
> i have tried more time but without success i do not know where i am wrong or if it is my smartphone too old...

There several methods to do this.  
Sometimes I'm using HttpCanary for Android. It intercept HTTP traffic and can export recorded packets into series of files. That is inconvient.  
But what better thing HttpCanary have is for HTTPS they suggest to use modifed version of **Parallel Space**. This modified version allows for certificates to be self-signed. In result you can use any MITM software. For example **mitmproxy** or **Charles Web Proxy**. I'm prefer latter because of GUI but it is paid app.

So basically I've setup Android for proxy via my PC where MITM software installed. For SSL traffic it replaces server certificates for own. For Android>7.0 you could not install CA from MITM proxy. So you need apps like **Parallel Space** which sits between Android and your target app like Netflix app and allows to use self-signed server sertificate.

I can send you mine incercepted session in **Charles Web Proxy** or **HAR** formats but not via Github.

**Comment by CastagnaIT** [comment]
thanks for sharing i will try again capture taffic another day

i have tried various combination to get `extrasFeed` but data without success,
always return empty data, usually means that could be limited to mobile endpoint :(

**Comment by tarhan** [comment]
Not sure. I've tried to repeat yesterday's request as CURL request. And it is still valid. So I think cookies does not contains any expired info.

**Comment by jakermx** [comment]
AFAIK....Navigation queries are done by lolomo, MSL is used to get Tittle playback relatad  stuff like URLS forvideo,audio and subs.

Comming soon option is a great enhanment,but  I think it could be nice to have external support forenhanced views and get extra info

**Comment by CastagnaIT** [comment]
> Not sure. I've tried to repeat yesterday's request as CURL request. And it is still valid. So I think cookies does not contains any expired info.

yes it is working on mobile endpoint
but i have tryed test with addon that use endpont of website (not the mobile endpoint),
and this not always have access to the same data

**Comment by jakermx** [comment]
> > Not sure. I've tried to repeat yesterday's request as CURL request. And it is still valid. So I think cookies does not contains any expired info.
> 
> yes it is working on mobile endpoint
> but i have tryed test with addon that use endpont of website (not the mobile endpoint),
> and this not always have access to the same data

Have you tried  using the geo located page, it could be the reasson of the non accurate results. ... maybe  because the news  are related to the Geo Zone.

**Comment by Sil3ntKn1ght** [comment]
> 
> 
> > Not sure. I've tried to repeat yesterday's request as CURL request. And it is still valid. So I think cookies does not contains any expired info.
> 
> yes it is working on mobile endpoint
> but i have tryed test with addon that use endpont of website (not the mobile endpoint),
> and this not always have access to the same data

---------------------------------------------------------------------------------------------------------------------------------------------
Add-on issues started today no updates nothing has changes on windows 10, kodi installed from download from there Kodi page, have tried fresh install from scratch with kodi netflix issue still there.

The add-on encountered the following error
Exception: KeyError
Error Details:
'Loco'

Last log entries 
2021-03-31 09:38:15.133 T:14704    INFO <general>: XBApplicationEx: destroying...
2021-03-31 09:38:15.497 T:14704    INFO <general>: AddOnLog: peripheral.joystick: Disabling joystick interface "directinput"
2021-03-31 09:38:15.497 T:14704    INFO <general>: AddOnLog: peripheral.joystick: Disabling joystick interface "xinput"
2021-03-31 09:38:15.497 T:14704    INFO <general>: ADDON: Dll Destroyed - Joystick Support
2021-03-31 09:38:15.498 T:14704    INFO <general>: unload skin
2021-03-31 09:38:15.527 T:14704    INFO <general>: Unloaded skin
2021-03-31 09:38:15.567 T:14704    INFO <general>: unload sections
2021-03-31 09:38:15.576 T:14704    INFO <general>: XBApplicationEx: application stopped!

**Comment by CastagnaIT** [comment]
"lolomo" is a way that Netflix uses to organise the data of the lists of tvshow/movies and his sub-lists etc..
sidenote, although there is still some reference on the website of "lolomo" is not used anymore,
now there is "LoCo" that is more o less similar

VPN topic here is not allowed here, the mentioned "Geo Zone" has nothing to do here

unless someone manages to circumvent the limits for accessing data through the endpoints we have (I have many doubts...)
i think the only way is get the data from "New & Popular" section of website

**Comment by tarhan** [comment]
> unless someone manages to circumvent the limits for accessing data through the endpoints we have (I have many doubts...)

So I need to find way to create valid requests with UserAgent used in add-on to coming soon endpoints? Can point in what class you are creating HTTPS navigation requests? In `Directory` in `resources/lib/navigation/directory.py`?

**Comment by CastagnaIT** [comment]
> So I need to find way to create valid requests with UserAgent used in add-on to coming soon endpoints? Can point in what class you are creating HTTPS navigation requests? In `Directory` in `resources/lib/navigation/directory.py`?

is a more complex because we need a more generic way to handle all menus,
the menu configuration is in globals.py
where is specified each menu what method have to call from `resources/lib/navigation/directory.py` path
then each method in `resources/lib/navigation/directory.py` requests the data from services/nfsession/directorybuilder/
where the requests are in the dir_path_requests.py
but you can not find a direct http request call, due to website complexity there are some data conversion, other checks etc..
then you will find only `self.nfsession.path_request` calls

if you have to do experiments with useragent or other endpoints
i think the "easy" way is add a new endpoint to endponts.py
and create a temporary customm method derived from `self.nfsession.path_request` method
then hijack a menu to try get your custom data

**Comment by tarhan** [comment]
ok. I will try it tomorrow.

**Comment by tarhan** [comment]
@CastagnaIT I've played with CURL request from intercepted requests. Even if I remove all HTTP headers except `Cookie` it will continue to work.  

But if I remove any of four cookies (`nfvdid`, `flwssn`, `NetflixId`, `SecureNetflixId`) then either I get erorr 403 or I receive JSON response with array items replaced with following:
```json
    {
      "message": "Unknown cause",
      "unauthorized": true
    }
```

**Comment by CastagnaIT** [comment]
this ^^^ is to say what? So what is the purpose?

**Comment by tarhan** [comment]
I've mentioned about it because for example in similar add-on for streaming service viki.com some HTTP headers are mandatory. You can have valid cookies with all authentication info filled but without some HTTP you will get 403 or 422 error.  
So I've mentioned that at least for Coming Soon section HTTP headers are not mandatory.

For you as expert of Netflix requests this is could be obvious but still `android.prod.ftl.netflix.com` endpoint could have different requirements than one you are using.

**Comment by jakermx** [comment]
To see ,headers, paramters and requests, I browse NF wit developer mode(F12) opened, then select Network Tab and there you go....  :)

**Comment by tarhan** [comment]
@jakermx You have misunderstood me I think. I know how to use Dev Tools of desktop browsers or MITM software.

Here I gave example of request made by Android Netflix client app. Maybe required HTTP Headers and auth differs from web version which @CastagnaIT used to.

**Comment by CastagnaIT** [comment]
i did a fast test with a python script
the android endpoint do not accept the cookies of website login, raise error

`
{"code":"Internal","message":"DnaClientError: failed DNA request: <!doctype html><html lang=\"en\"><head><title>HTTP Status 403 – Forbidden</title><style type=\"text/css\">body {font-family:Tahoma,Arial,sans-serif;} h1, h2, h3, b {color:white;background-color:#525D76;} h1 {font-size:22px;} h2 {font-size:16px;} h3 {font-size:14px;} p {font-size:12px;} a {color:black;} .line {height:1px;background-color:#525D76;border:none;}</style></head><body><h1>HTTP Status 403 – Forbidden</h1></body></html>"}
`

i do not think that can be used

**Comment by tarhan** [comment]
> the android endpoint do not accept the cookies of website login, raise error

Yeah. Done same.  
I've cutted request to shorter version:
```
curl -H 'Host: android.prod.ftl.netflix.com' -H 'Cookie: ......' --compressed 'https://android.prod.ftl.netflix.com/nq/androidui/samurai/~7.95.0/api?path=%5B%22extrasFeed%22%2C%7B%22from%22%3A0%2C%22to%22%3A49%7D%2C%22video%22%2C%5B%22summary%22%5D%5D'
```

I.e. JSON Graph request only requests path `["extrasFeed",{"from":0,"to":49},"video",["summary"]]` without any additional HTTP headers. And it works.
Cookies to Android endpoint contains following entries:  `nfvdid`, `flwssn`, `NetflixId`, `SecureNetflixId`.  
If I replace `nfvdid` with web version all continue to works.  
If I replace `SecureNetflixId` with web version I receive empty 40 entries in json graph response.  
If I replace `NetflixId` with web version I get your error (403).

`NetflixId` itself is URL encoded query. Which contains `v`, `ct`, `bt`, `ch`, `mac` entries. `ch` entry is same for Android and web version.

**Comment by CastagnaIT** [comment]
Issue closed, FR implemented in the new release

**Comment by tarhan** [comment]
Wonderful
